### PR TITLE
Rework the API to allow message batching

### PR DESCRIPTION
**Pull request by nagisa**
In general these changes have occured:

1. `publish` no longer returns a `String` with a publish ID;
2. Major APIs have been changed to be more build-ery:
    * `Message` is now constructable stand-alone and is nothing more than just a bundle of data with no stateful references to any Hedwig instance;
     * `Message` now has methods to build it up further by specifying an UUID and headers;
     * Sending messages now occurs through `HedwigPublishBuilder`;
     * The actual publishers receive a vector of `(str, MessageSchema)` with first field being the topic and the 2nd field being the validated and encoded message, they then publish all the messages at once, batching the remote API calls as much as possible (e.g. google pubsub can only batch messages to a single topic, so it will sort the list and then publish in batches).

### CODE REVIEW

**Comment by nagisa** [comment]
Do not merge yet, I’m going to evaluate this API.

**Comment by nagisa** [comment]
@maroux this is ready for a review I think.

File: src/lib.rs [hunk 1]
```diff
@@ -499,153 +485,217 @@ where
...
+
+    /// Custom headers. This may be used to track request_id, for example.
+    pub headers: Headers,
+}
+
+/// A validated message.
+///
+/// This data type is the schema or the json messages being sent over the wire.
+#[derive(Debug, Serialize)]
+pub struct MessageSchema {
```
Comment (maroux) [inline src/lib.rs]: This doesn't need to be different from `Message` does it?

File: src/lib.rs [hunk 2]
```diff
@@ -499,153 +485,217 @@ where
...
+            schema,
+            format_version,
+            data: serde_json::to_value(self.data)?,
+        })
     }
 }
 
+/// Additional metadata associated with a message
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub struct MetadataSchema {
```
Comment (maroux) [inline src/lib.rs]: Same for this

File: src/lib.rs [hunk 3]
```diff
@@ -154,44 +154,38 @@ impl From<SchemaError> for HedwigError {
...
-/// favorite cloud platform\>.
+/// Message publishers.
+///
+/// This is used to interface with arbitrary pub-sub services.
 pub trait Publisher {
     /// Publish a Hedwig message.
-    fn publish<D, T>(&self, message: Message<D, T>) -> Result<String, PublishError>
-    where
-        D: Serialize;
+    fn publish(&self, message: Vec<(&'static str, MessageSchema)>) -> Result<(), PublishError>;
```
Comment (maroux) [inline src/lib.rs]: Needs document for the vector item..

File: src/lib.rs [hunk 4]
```diff
@@ -254,61 +249,70 @@ impl GooglePublisher {
             google_cloud_project,
         })
     }
+
+    fn publish_batch(
+        &self,
+        topic: &str,
+        batch: Vec<pubsub1::PubsubMessage>,
+    ) -> Result<(), PublishError> {
```
Comment (maroux) [inline src/lib.rs]: return message ids?

File: src/lib.rs [hunk 5]
```diff
@@ -499,153 +485,217 @@ where
...
-    /// * data_schema_version - Version of the data object
-    /// * data - The message data - must be serializable
-    pub fn message<D>(
-        &self,
-        data_type: T,
-        data_schema_version: Version,
-        data: D,
-    ) -> Result<Message<D, T>, MessageError>
+    /// Create a builder for a batch publish construction.
+    pub fn start_publish(&self) -> HedwigPublishBuilder<T, P> {
```
Comment (maroux) [inline src/lib.rs]: Should also mirror `HedwigPublisherBuilder::publish` method here so it doesn't force batch publish API even when publishing a single message.

File: src/lib.rs (hunk 1 above)
Comment (nagisa) [reply src/lib.rs]: The reason why they are separate is to separate the concerns these types handle. The `Message` is a builder whereas `MessageSchema` is the thing being serialized. Such separation aids maintainability of this code as `MessageSchema` should never change unless hedwig protocol changes, whereas it is perfectly fine for `Message` to change if we wanted to e.g. do something special about some extra data.

This also allows for validation of the message contents to not be spread out all over the place – previously the message contents were validated piecewise across the whole publish flow and the types did not reflect the state of validation at all.

Now the `Publisher::publish` takes the `MessageSchema` and so it knows that the argument it is receiving has been validated (and the data has been encoded to JSON) already.

Comment (maroux) [reply src/lib.rs]: Well, I suppose my point was that the class is simple enough that you don't need to separate out the transport from data packet. I agree that the validation should be more "centralized". For example, this class could be converted into an enum with state machine that transitions from "creating a message" to "validated message". After validation no changes can be made to the data.

As it's written currently, the class name isn't very descriptive. Can we call this `ValidatedMessage` or `MessageTransport` .. something along those lines?

Comment (nagisa) [reply src/lib.rs]: > For example, this class could be converted into an enum with state machine that transitions from "creating a message" to "validated message". After validation no changes can be made to the data.

I strongly disagree with that sentiment. Having multiple types, however similar they are, to represent different type states (and thus to enforce correct usage of them at the compile time) is a well known and widely used pattern. Using anything less usually means the invariants can only be enforced at run time.

---

The name here comes from the fact that this type represents the hedwig on-wire data 1:1. Thus "schema". I don’t have a strong opinion, though, so whatever, I’ll name it a validated message instead.

Comment (nagisa) [reply src/lib.rs]: FWIW we could do with just a single type while still retaining the compile-time guarantees, but in that case this library will want some more rework of its overall structure (esp. pertaining to the `Validator` and `Publisher`).

Comment (maroux) [reply src/lib.rs]: I mean its still different types ([this](https://hoverbear.org/2016/10/12/rust-state-machine-pattern/) is the pattern I'm referring to), but this is good enough.

**Comment by nagisa** [comment]
Any remaining/pressing outstanding concerns?

File: src/lib.rs [hunk 6]
```diff
@@ -499,153 +504,228 @@ where
...
-    /// * data_schema_version - Version of the data object
-    /// * data - The message data - must be serializable
-    pub fn message<D>(
-        &self,
-        data_type: T,
-        data_schema_version: Version,
-        data: D,
-    ) -> Result<Message<D, T>, MessageError>
+    /// Equivalent to `hedwig.build_publish().message(msg)?.publish()`.
+    pub fn publish<D>(&self, msg: Message<D, T>) -> Result<P::MessageIds, PublishError>
```
Comment (maroux) [inline src/lib.rs]: Return value should contain a single message id rather than a list.

**Pull request merged**